### PR TITLE
Fabo/1884 bring back e2e tests

### PR DESCRIPTION
**Pull request by faboweb**
Enables #1884

**Description:**

Thank you! 🚀

---

For contributor:

- [ ] Added entries in `CHANGELOG.md` with issue # and GitHub username
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by faboweb** [comment]
Things still need to be done:
- Prevent side-effects between runs
- Provide test with rich account credentials

**Comment by faboweb** [comment]
locally working

@sabau on circleci it can't find the gaia binaries, could you help out?

File: package.json [hunk 1]
```diff
@@ -27,7 +27,8 @@
     "lint:fix": "eslint --fix -f ./node_modules/eslint-friendly-formatter \"{,**/}*.{js,vue}\" && stylelint \"**/*.{css,vue}\" --fix",
     "test": "yarn run lint && yarn test:unit && yarn test:e2e",
     "test:unit": "cross-env NODE_ENV=testing jest --maxWorkers=2",
-    "test:e2e": "cross-env yarn run pack && tape \"${TEST:-test/e2e/*.js}\"",
+    "test:e2e": "(kill $(lsof -t -i:9515) || true) && nightwatch --config ./test/e2e/nightwatch.json",
```
Comment (faboweb) [inline package.json]: ugly and not cross platform. but I couldn't fix it in nightwatch :/

File: .circleci/config.yml [hunk 2]
```diff
@@ -188,23 +188,23 @@ jobs:
...
-  #          name: Test
-  #          command: yarn run test:e2e
-  #          no_output_timeout: 120
-  #      - store_artifacts:
-  #          path: testArtifacts
-  #          when: on_fail
+  testE2e:
+    executor: web
+    environment:
+      - BINARY_PATH: "/tmp/voyager/linux_amd64/gaiacli"
```
Comment (faboweb) [inline .circleci/config.yml]: `/tmp/voyager` is the workspace, this is why the gaia files are attached there. we need to clean the paths up.

**Comment by faboweb** [comment]
> i would feel better if @sabau reviewed this too

likewise :)

File: tasks/build/Gaia/build.sh [hunk 3]
```diff
@@ -23,11 +23,14 @@ for GOOS in $platforms; do
   make get_vendor_deps
   make install
 
-  if [[ "$GOOS" == "linux" ]]; then
+  if [ "$GOOS" = "linux" ]; then
```
Comment (sabau) [inline tasks/build/Gaia/build.sh]: there is an advantage in switching to the legacy `[` against `[[`? it should be more portable but I think this will anyway run on up to date machines / containers

Comment (faboweb) [reply tasks/build/Gaia/build.sh]: there were issues with the old notation `[[ "$GOOS" == "linux" ]]` on circleci
the whole build process didn't run so we didn't see that this actually would have failed

File: tasks/build/Gaia/build.sh [hunk 4]
```diff
@@ -23,11 +23,14 @@ for GOOS in $platforms; do
...
     # Give the Linux binaries a prefix to make them like the others.
     mkdir --parents $TARGET/linux_amd64
     cp /go/bin/gaia* $TARGET/linux_amd64/
   fi
 done
 
-cp --recursive /go/bin/*_amd64 $TARGET/
+# copy build files to target dir
+if stat -t /go/bin/*_amd64 >/dev/null 2>&1; then
```
Comment (sabau) [inline tasks/build/Gaia/build.sh]: I think I would prefer the command to error if nothing has been produced

File: test/e2e/nightwatch.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    "custom_commands_path": null,
+    "custom_assertions_path": null,
+    "page_objects_path": null,
+    "globals_path": "./globals.js",
+    "disable_colors": false,
+    "test_workers": false,
+    "webdriver": {
+        "start_process": false,
+        "host": "localhost",
+        "port": 8080
```
Comment (sabau) [inline test/e2e/nightwatch.json]: is not conflicting with our proxy?

File: tasks/build/local/helper.js [hunk 6]
```diff
@@ -55,27 +55,20 @@ const saveVersion = nodeHome => {
...
-          (await makeValidator(
-            mainAccountSignInfo,
-            nodes[i].home,
-            nodes[i].cliHome,
-            nodes[i].moniker,
-            chainId
-          ))
-      })()
-    )
+    // start secondary nodes and connect to node one
```
[the remark this answers is not in the document]
Comment (faboweb) [reply tasks/build/local/helper.js]: starting in parallel produced issues with signing :/ time difference is very low

File: tasks/gaia.js [hunk 7]
```diff
@@ -166,14 +166,15 @@ async function sendTokens(
   toAddress,
   chainId
 ) {
+  await sleep(500)
```
Comment (sabau) [inline tasks/gaia.js]: Just wondering: why was this necessary before sending the money?

File: tasks/build/Gaia/build.sh (hunk 4 above)
Comment (faboweb) [reply tasks/build/Gaia/build.sh]: Should we exit before if we build only for linux?

File: tasks/build/Gaia/build.sh (hunk 3 above)
Comment (faboweb) [reply tasks/build/Gaia/build.sh]: adding: open to trying this again with you

File: tasks/gaia.js [hunk 8]
```diff
@@ -275,13 +276,25 @@ function makeExecWithInputs(command, inputs = [], json = true) {
...
   })
 }
 
+/*
+
+    mnemonic: `release endorse scale across absurd trouble climb unaware actor elite fantasy chair license word rare length business kiss smoke tackle report february bid ginger`,
+    address: `cosmos1ek9cd8ewgxg9w5xllq9um0uf4aaxaruvcw4v9e`,
+    */
+function sendMoneyToFixedAccounts(mainAccountSignInfo, chainId) {
+  sendTokens(mainAccountSignInfo, `${100 * 10e6}stake`, `cosmos1ek9cd8ewgxg9w5xllq9um0uf4aaxaruvcw4v9e`, chainId)
```
Comment (sabau) [inline tasks/gaia.js]: I would just move `cosmos1ek9cd8ewgxg9w5xllq9um0uf4aaxaruvcw4v9e` to a variable, I've seen it's copy pasted in a couple of places

File: .circleci/config.yml [hunk 9]
```diff
@@ -188,23 +188,23 @@ jobs:
...
+      - NODE_BINARY_PATH: "/tmp/voyager/linux_amd64/gaiad"
+    steps:
+      - checkout
+      - attach_workspace:
+          at: *WORKSPACE
+      - yarn-install
+      - run:
+          name: Test
+          command: yarn serve & yarn run test:e2e
```
Comment (sabau) [inline .circleci/config.yml]: Do we need to use a server for static files?

File: package.json [hunk 10]
```diff
@@ -21,13 +21,14 @@
     "build:ui": "cross-env NODE_ENV=production webpack --colors --config webpack.renderer.config.js --mode=production --profile --json > app/dist/build.stats.json",
     "start": "concurrently -k \"yarn frontend\" \"yarn backend\"",
     "start:new": "concurrently -k \"yarn frontend\" \"yarn backend:new\"",
+    "serve": "http-server -p 8081 ./app/dist",
```
Comment (sabau) [inline package.json]: wouldn't be ok to use our yarn frontend while developing? and on circle run against the artifacts

File: package.json (hunk 1 above)
Comment (sabau) [reply package.json]: A nightwatch side project owner said he will fix it in a week (on january 2016)(https://github.com/mucsi96/nightwatch-cucumber/issues/12#issuecomment-174865979) let's keep this dirty stuff XD

File: tasks/build/Gaia/build.sh (hunk 3 above)
Comment (sabau) [reply tasks/build/Gaia/build.sh]: That's funny, the `[[` should be the current notation, ok let's stick to the legacy and see in another PR what docker image need to be updated

File: tasks/build/Gaia/build.sh (hunk 4 above)
Comment (sabau) [reply tasks/build/Gaia/build.sh]: Ah probably I tested only a subset here, let's keep it simple, we will anyway change this again to remove `darwin` from our package.json

File: .circleci/config.yml (hunk 9 above)
Comment (faboweb) [reply .circleci/config.yml]: I don't know exactly. will need to recheck.

Comment (sabau) [reply .circleci/config.yml]: but I think for now let's just use this working version, you probably invested already enough time on this PR right? we'll have to touch it to include browserstack so let's move this when will be vital to do so :)

Comment (sabau) [reply .circleci/config.yml]: I was trying to prepare the path to have a ~1 line change for browserstack 🤣

File: tasks/gaia.js (hunk 7 above)
Comment (faboweb) [reply tasks/gaia.js]: I think this was due to the parallel signing issues. I will test this again. Nice find.
